.github/workflows/build-test.yaml

```diff
@@ -206,36 +206,92 @@ jobs:
       - name: Run K3s and master in it
         run: tests/run-master-in-k3s.sh local/freeipa-server:${{ matrix.os }} freeipa-server-${{ matrix.os }}.tar
 
+  push-after-success-precondition:
+    name: Check delivery to registries
+    needs: [ test-docker, test-podman, test-rootless-podman, test-upgrade, test-k3s ]
+    runs-on: ubuntu-20.04
+    env:
+      REGISTRY_TARGET_LIST: "${{ secrets.REGISTRY_TARGET_LIST }}"
+      REGISTRY_CREDENTIALS_FILE: "${{ secrets.REGISTRY_CREDENTIALS_FILE }}"
+    steps:
+      - name: Check secrets
+        id: check
+        run: |
+          [ "${REGISTRY_CREDENTIALS_FILE}" == "" ] && {
+            echo "REGISTRY_CREDENTIALS_FILE is empty"
+            echo "::set-output name=are-valid-credentials::false"
+            exit 0
+          }
+          [ "${REGISTRY_TARGET_LIST}" == "" ] && {
+            echo "REGISTRY_TARGET_LIST is empty"
+            echo "::set-output name=are-valid-credentials::false"
+            exit 0
+          }
+          IFS=":" read -a TARGET_LIST <<< "${REGISTRY_TARGET_LIST}"
+          [ ${#TARGET_LIST[@]} -eq 0 ] && {
+            echo "TARGET_LIST is empty; Check your REGISTRY_TARGET_LIST secret in your repository settings"
+            echo "::set-output name=are-valid-credentials::false"
+            exit 0
+          }
+          echo "::set-output name=are-valid-credentials::true"
+    outputs:
+      are-valid-credentials: ${{ steps.check.outputs.are-valid-credentials }}
+
   push-after-success:
-    name: Push images to Docker Hub
+    name: Delivery images to registries
     runs-on: ubuntu-20.04
-    needs: [ test-docker, test-podman, test-rootless-podman, test-upgrade, test-k3s ]
-    if: github.event_name != 'pull_request' && github.repository == 'freeipa/freeipa-container' && github.ref == 'refs/heads/master'
+    needs: [ push-after-success-precondition ]
+    if: needs.push-after-success-precondition.outputs.are-valid-credentials == 'true' && github.event_name != 'pull_request' && github.repository == 'freeipa/freeipa-container' && github.ref == 'refs/heads/master'
```

Comment:
I'm not sure I follow the intent. I thought the goal per https://github.com/freeipa/freeipa-container/issues/356 and https://github.com/freeipa/freeipa-container/issues/363 was for developers to be able to push from their forks to their registry targets. But here the push only happens for `github.repository == 'freeipa/freeipa-container' && github.ref == 'refs/heads/master'` ... but for that we could have just hardcoded the targets and wouldn't need to make things customizable.

Reply:
I have thought the following change:

1. I would add the below to the **Check settings** step, before the current behavior:
```yaml
steps:
      - name: Check settings
        id: check
        run: |
          # Check the branch is allowed to push the image
          read -a BRANCH_LIST <<< "${REGISTRY_ALLOWED_BRANCH_LIST}"
          if [ ${#BRANCH_LIST[@]} -eq 0 ]; then
            is_branch_allowed=0
          else
            is_branch_allowed=0
            for branch in ${BRANCH_LIST[@]}; do
              [ "refs/heads/${branch}" == ${{ github.branch }} ] && {
                is_branch_allowed=1
                break
              }
            done
          fi
          # Check the credentials are filled
          if [ "${REGISTRY_CREDENTIALS_FILE}" != "" ]; then
            are_credentials_present=1
          else
            echo "REGISTRY_CREDENTIALS_FILE secret is empty or does not exist"
            are_credentials_present=0
          fi
          # Check the target list
          if [ "${REGISTRY_TARGET_LIST}" == "" ]; then
            echo "REGISTRY_TARGET_LIST secret is empty or does not exist"
            has_target_list=0
          else
            IFS=":" read -a TARGET_LIST <<< "${REGISTRY_TARGET_LIST}"
            if [ ${#TARGET_LIST[@]} -eq 0 ]; then
              echo "Check REGISTRY_TARGET_LIST secret in your repository settings"
              has_target_list=0
            else
              has_target_list=1
            fi
          fi
          # Set the output value
          if [ "${is_branch_allowed}" -eq 1 ] \
             && [ "${are_credentials_present}" -eq 1 ] \
             && [ "${has_target_list}" -eq 1 ]; then
            echo "::set-output name=is-allowed-to-push::true"
          else
            echo "::set-output name=is-allowed-to-push::false"
          fi
```
2. I would change the **if** by this new one:
```yaml
    if: needs.push-after-success-precondition.outputs.is-allowed-to-push == 'true' && github.event_name != 'pull_request'
```

1) The secrets are per repository, so the parent repository and the forked repository are working with different secrets, which means that credentials for pushing to the repository and list of repositories are customized per repository and one repository can not access the registry the other is pushing images (unless we provide the necessary settings for it, this is, same credentials and target repositories).

2) In regarding with the branch restriction, I would use the variable `REGISTRY_ALLOWED_BRANCH_LIST` as the list of branches that can deliver images, so that each repository has their own list:

```raw
REGISTRY_ALLOWED_BRANCH_LIST="master 363-add-support-for-multiple-registries"
```
----

All the above would allow to push the customized branches, and that configuration would be different per repository. When I fork the repository I would have to set up the secrets properly for allowing the pipeline to delivery into my customized registry (or registries), and only the specified branches would delivery a container image customized by repository; the credentials should be set up into the forked repository for allowing the delivery too. If the credentials are not set up, the job is just skipped.

If you agree with the above, I guess the change should be into the #357 instead of this PR, but that is just a matter of where to encapsulate the change.

I have created a child branch for testing the above:
https://github.com/avisiedo/freeipa-container/blob/363-b-add-support-for-multiple-registries/.github/workflows/build-test.yaml#L194

----

What do you think? Let me know your thoughts.

Reply:
I think it should not be list of allowed branches and list of targets. It should be list of targets and for each of them list of branches.

However, I believe we need to go back to the requirements stage and drawing board. What is the goal here? How will the users (people pushing or merging to their forks) use this feature?

If you fork https://github.com/freeipa/freeipa-container/ and you don't do any change to the code, you can just as well use the images from the registries to which we send them. If you do some changes to those branches or you want to push from different branch than master, you can tweak .github/workflows/build-test.yaml as well.

Frankly, I'm inclined to split the effort, just do a change which will allow us to automate pushing to Quay besides Docker Hub, with no support for pushes from forks, and deal with those separately. I'll prepare a prototype to that effect.

Reply:
Here's change enabling push to multiple registries: https://github.com/adelton/freeipa-container/commit/153241f63dec5e1351c6b919e97f6abf9235c0fc. The same set of tags is pushed to all targets, it is not customizable at this point. It is expected to only run on https://github.com/freeipa/freeipa-container/'s master, no support for pushes from forks or other branches.

I've tested the setup with https://github.com/adelton/freeipa-container/commit/57fbd5ae0d9c9541865c7e41efcd10435ae176ea, forcing it to run on my branch, and pushes to https://quay.io/repository/adelton/freeipa-container-test?tab=tags and https://hub.docker.com/r/adelton/freeipa-server/tags?page=1&ordering=last_updated passed fine, here is the log: https://github.com/adelton/freeipa-container/runs/1887384540.

Reply:
The change for master shown as https://github.com/freeipa/freeipa-container/pull/376.

Reply:
I've now merged https://github.com/freeipa/freeipa-container/pull/376 and the images got pushed both to Quay and Docker Hub.

That created conflict for this pull request of your but I'd really like to get clarity of the desired behaviour for potential pushes from forked repository before continuing review.

Reply:
Specifically, now that we will (hopefully) be pushing to Quay with each push to master and then three times a week, so that you don't hit rate limits on Docker Hub -- what specifically do you guys need in addition to that, for your effort?